### PR TITLE
Better server exit code

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The server's exit code will correspond to the exception code. For example, if the server cannot start due to memory limit, it will exit with the code 241 = MEMORY_LIMIT_EXCEEDED. In previous versions, the exit code for exceptions was always 70 = Poco::Util::ExitCode::EXIT_SOFTWARE.

```
int Application::run()
{
	int rc = EXIT_CONFIG;
	initialize(*this);

	try
	{
		rc = EXIT_SOFTWARE;
		rc = main(_unprocessedArgs);
	}
	catch (Poco::Exception& exc)
	{
		logger().log(exc);
	}
	catch (std::exception& exc)
	{
		logger().error(exc.what());
	}
	catch (...)
	{
		logger().fatal("system exception");
	}

	uninitialize();
	return rc;
}
```

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit bd4ffec361234da945136cd33039e756085dc8ce with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57153/bd4ffec361234da945136cd33039e756085dc8ce/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6972013744">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57153/bd4ffec361234da945136cd33039e756085dc8ce/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by thevar1able — APPROVED** [review]

**Comment by Algunenano** [comment]
> The server's exit code will correspond to the exception code.

Just a note. IIRC in Posix this is limited to 256, so if you return any bigger number you will get a "truncated" error code, e.g. `CANNOT_LOAD_CONFIG` -> 347 -> 347 - 256 -> 91

This means that some error codes that are multiple of 256 will return 0:
```
M(256, PARTITION_ALREADY_EXISTS) \
M(512, SET_NON_GRANTED_ROLE) \
```

We should avoid that

**Comment by alexey-milovidov** [comment]
Good point, and we already have this logic in Client and Local apps.

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -1970,7 +1970,8 @@ catch (...)
 {
     /// Poco does not provide stacktrace.
     tryLogCurrentException("Application");
-    throw;
+    auto code = getCurrentExceptionCode();
+    return code ? code : -1;
```
Comment (Algunenano) [inline programs/server/Server.cpp:1974]: This doesn't seem right. code is an int, so multiples of 256 are != 0 and the error code won't be changed, right?

```
catch (...)
{
    /// Poco does not provide stacktrace.
    tryLogCurrentException("Application");
    auto code = 256;
    return code ? code : -1;
}
```
Exits the program with 0 still

**Pull request merged**